### PR TITLE
NtOS Emojipedia - PDA Emoji Messaging For All

### LINKED ISSUE #67623: there's no more emojipedia on mime pdas
## Testmerges:
The following pull requests are currently test merged:
[#65791: 'Optimizes Runechat' by Kylerace at commit 775d002c97](https://www.github.com/tgstation/tgstation/pull/65791)
[#67385: 'Reduces the point reduction on traitor objectives significantly' by Watermelon914 at commit 9a9f35340b](https://www.github.com/tgstation/tgstation/pull/67385)
[#67116: 'TGUI Say: Upgrades chat input with modern features' by jlsnow301 at commit 9edb8d8709](https://www.github.com/tgstation/tgstation/pull/67116)
[#67615: 'Fix Dynamic midrounds spawning heavies significantly earlier than they're supposed to' by Mothblocks at commit 28b5169054](https://www.github.com/tgstation/tgstation/pull/67615)

## Reproduction:
pretty sure it's been like this since the tablet change but noone's noticed
used to be a big list of emojis and the code to use them in the mime pda but that functionality has not carried over to tablets

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

Firstly, apologies for any technical/visual grossities. This is the first full interface I've created using TGUI, and although I tried multiple ways to make it look better, I just wasn't ever able to get it to work (lack of knowledge). I'm also probably missing something big, so apologies for that as well.

Anyways, this PR re-adds Emojipedia (first added in #47177), a tablet function that was lost when we swapped from HTML PDAs to TGUI Tablets. It wasn't a spectacular program, best I could tell in my research all it did was spit out all of the potential emotes to the user's chat with the associated word. This is a bit better than that since it's an actual program application.

One thing I discovered during this PR is that we have three "animated" emotes- however they do not work at all in TGChat, and they make the program look ugly. I pruned down these emotes to only contain one frame in their icon_state, because they just do not work anymore in our current framework.

After some discussion in the PR comments below, I also change the functionality in this PR to allow _everyone_ to send PDA messages! Huzzah! As a hold-over, I think I'll let Curator and Mime start with it shift-start as a reminder of where they once came from (and also because the mime _needs_ it).

## Why It's Good For The Game

image

image

image

Fixes #67623. This was dropped off for effectively no reason, and it's good to keep functionality going.

## Changelog

:cl: san7890, ZephyrTFA
fix: Emojipedia has been restored to tablets! Look for it in your local Nanotrasen Software Hub.
balance: Everyone is now able to send Emojis via PDA Messaging! It's no longer restricted to just mime and curators, so enjoy sending the classics such as :ai:, :1997:, :pslime:, and more!
/:cl:

### CODE REVIEW

**Comment by ZephyrTFA** [comment]
Think you could make it a grid of the icons and you can hover to see what the key is instead?
I feel like it would look a lot nicer

**Comment by san7890** [comment]
> Think you could make it a grid of the icons and you can hover to see what the key is instead? I feel like it would look a lot nicer

I really, really wanted to try making a grid, something like:
image

(imagine every cell has text)

But I tried finagling with TGUI and I just could not figure it out.

**Comment by ZephyrTFA** [comment]
> 

let me see if I can make a PR on your branch for this

**Comment by san7890** [comment]
> let me see if I can make a PR on your branch for this

I have skill of noob, thank you so much. Feel free to ping me on Discord `@san7890#7890` if you make the PR and I'm not paying attention or something.

File: code/modules/modular_computers/file_system/programs/emojipedia.dm [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/// A tablet app that lets anyone see all the valid emoji they can send via a PDA message (and even OOC!)
+/datum/computer_file/program/emojipedia
+	filename = "emojipedia"
+	filedesc = "EmojiPedia"
+	category = PROGRAM_CATEGORY_CREW // we want everyone to be able to access this application, since they can use it as a reference for posting in OOC (only mimes and curators can send emoji via PDAs right now)
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm:5]: i dont like this thought process regardless

[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm:5]: frankly we need to rethink emojis being mime/curator only. i have just been sending kaomojis in the meantime

[the remark this answers is not in the document]
Comment (san7890) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm:5]: I can expand emoji-sending to everyone on the station. I thought it was silly too when I discovered it, and it's not documented anywhere player-facing.

[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm:5]: That seems ok I suppose

[the remark this answers is not in the document]
Comment (san7890) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm]: PR updated to allow anyone to message emojis. This is from an Assistant's PDA:

image

File: tgui/packages/tgui/interfaces/NtosEmojipedia.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import { useBackend, useSharedState } from '../backend';
+import { Box, Button, Input, Section } from '../components';
+import { NtosWindow } from '../layouts';
+import '../styles/interfaces/Emojipedia.scss';
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NtosEmojipedia.js:4]: Import it via main.scss file for consistency and faster compile times.

File: tgui/packages/tgui/interfaces/NtosEmojipedia.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+              />
+            </>
+          }
+          display="grid"
+          box-sizing="border-box"
+          margin="0"
+          padding="0"
+          grid-template-columns="repeat(auto-fill, 32px)"
+          grid-template-rows="repeat(auto-fill, 32px)"
+          grid-gap="4em">
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NtosEmojipedia.js:46]: These properties, starting with display, aren't supposed to be working on Box and should better stay in CSS, and CSS grid support in IE11 is a little bad, keep that in mind, as long as it works as you expect.

File: code/modules/modular_computers/file_system/programs/emojipedia.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm]: this is fucked up bruh

File: code/modules/modular_computers/file_system/programs/emojipedia.dm [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	program_icon = "icons"
+	/// Store the list of potential emojis here.
+	var/static/list/emoji_list = icon_states(icon(EMOJI_SET))
+
+/datum/computer_file/program/emojipedia/New()
+	. = ..()
+	// Sort the emoji list so it's easier to find things and we don't have to keep sorting on ui_data since the number of emojis can not change in-game.
+	emoji_list = sortTim(emoji_list, /proc/cmp_text_asc)
+
+/datum/computer_file/program/emojipedia/ui_data(mob_user)
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm:19]: Static data would be better than ui_data here, but constant data would be even better

At the very least if this is moved to static data, cache the result of this

File: code/modules/modular_computers/file_system/programs/emojipedia.dm [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	var/list/data = get_header_data()
+	for(var/emoji in emoji_list)
+		data["emoji_list"] += list(list(
+			"name" = emoji,
+			"icon64" = icon2base64(icon(EMOJI_SET, emoji)),
```
Comment (Mothblocks) [inline code/modules/modular_computers/file_system/programs/emojipedia.dm:24]: You should make this a spritesheet instead

Comment (Mothblocks) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm:24]: `icon2base64` payloads are kinda huge

File: code/modules/modular_computers/file_system/programs/emojipedia.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (san7890) [reply code/modules/modular_computers/file_system/programs/emojipedia.dm]: I'm not entirely sure how to cache the result or use constant data here.

**Comment by san7890** [comment]
I don't know if it's a limitation of spritesheets or a limitation of how we handle spritesheets via CSS, but basically any icon_states that were just raw numbers would break like such:
image

With `pensive`'s icon being the last icon in the DMI. I decided that number emojis kinda sucked anyways, so I just removed those and added a unit test to lint against it.

image

:1: going to `:exclamation:`
:1997: going to `:malf:`

image

File: code/modules/unit_tests/_unit_tests.dm [hunk 6]
```diff
@@ -191,6 +191,7 @@
 #include "timer_sanity.dm"
 #include "traitor.dm"
 #include "unit_test.dm"
+#include "verify_emoji_names.dm"
 #include "verify_config_tags.dm"
```
Comment (Mothblocks) [inline code/modules/unit_tests/_unit_tests.dm:195]: ABC order

File: code/modules/unit_tests/verify_emoji_names.dm [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/// Apparently, spritesheets (or maybe how the CSS backend works) do not respond well to icon_state names that are just pure numbers (which was a behavior in emoji.dmi).
+/// In case we add more emoji, let's just make sure that we don't have any pure numbers in the emoji.dmi file if we ever add more.
+/datum/unit_test/verify_emoji_names
+	priority = 0 // san7890 remove this
```
Comment (Mothblocks) [inline code/modules/unit_tests/verify_emoji_names.dm:4]: just use TEST_FOCUS bro

Comment (san7890) [reply code/modules/unit_tests/verify_emoji_names.dm:4]: I SWORE I REMOVED THIS

**Pull request merged**